File: python-code-modified/boundary_conditions.py
from helper import *
from parameters import SimulationParametes as parameters
from parameters import Constants as constants
from parameters import BoundaryInfo as boundary_info


def get_bc_function(flag):

    if flag == "wall":
        return apply_non_slip_bc
    elif flag == "mov_wall":
        return apply_mov_wall_bc
    else:
        return skip_boundary


def skip_boundary(component,
                  coordinate,
                  population,
                  velocity,
                  density):
    pass


def apply_non_slip_bc(component,
                      coordinate,
                      population,
                      velocity,
                      density):

    num_lattices = parameters.num_lattices
    num_directions = parameters.discretization

    # Compute inverse coordinate
    i = -1 * constants.coords[component]
    j = -1 * constants.coords[component + num_directions]

    # Find inverse index
    inverse_component = constants.inverse_indices[component]

    # compute the gap to the target fluid lattice
    shift = get_index(i, j, dim=1)

    self_index = (coordinate) + component * num_lattices
    neightbour_index = (coordinate + shift) + inverse_component * num_lattices

    population[neightbour_index] = population[self_index]


def apply_mov_wall_bc(component,
                      coordinate,
                      population,
                      velocity,
                      density):

    num_directions = parameters.discretization
    num_lattices = parameters.num_lattices

    # Compute inverse coordinate
    i = -1 * constants.coords[component]
    j = -1 * constants.coords[component + num_directions]

    # Find inverse index
    inverse_component = constants.inverse_indices[component]

    # compute the dot product Cinv.T * Uwall
    dot_product_CU = constants.const_1 * (i * boundary_info.wall_velocity[0] +
                                          j * boundary_info.wall_velocity[1])

    # compute the shift to the target fluid lattice
    shift = get_index(i, j, dim=1)

    self_index = (coordinate) + component * num_lattices
    neightbour_index = (coordinate + shift) + inverse_component * num_lattices

    scalar_neighbour_index = coordinate + shift

    population[neightbour_index] = (population[self_index] +
                                    (2.0 * constants.weights[component] *
                                     density[scalar_neighbour_index] *
                                     dot_product_CU))

def apply_inflow_bc(component,
                    coordinate,
                    population,
                    velocity,
                    density):

    num_directions = parameters.discretization
    num_lattices = parameters.num_lattices

    # Compute inverse coordinate
    i = -1 * constants.coords[component]
    j = -1 * constants.coords[component + num_directions]

    # Find inverse index
    inverse_component = constants.inverse_indices[component]

    # compute the shift to the target fluid lattice
    shift = get_index(i, j, dim=1)

    self_index = (coordinate) + component * num_lattices
    neightbour_index = (coordinate + shift) + inverse_component * num_lattices

    dot_product_UU = (boundary_info.velocity_inflow[0] * boundary_info.velocity_inflow[0] +
                      boundary_info.velocity_inflow[1] * boundary_info.velocity_inflow[1])

    dot_product_CU = i * boundary_info.velocity_inflow[0] + j * boundary_info.velocity_inflow[1]


    velocity_expansion = (constants.const_1 * dot_product_CU +
                          constants.const_2 * dot_product_CU * dot_product_CU -
                          constants.const_3 * dot_product_UU +
                          1.0)

    # Compute equilibrium distribution
    equilibrium = constants.weights[inverse_component] * boundary_info.density_inflow * velocity_expansion

    population[neightbour_index] = equilibrium


def apply_outflow_bc(component,
                     coordinate,
                     population,
                     velocity,
                     density):


    num_directions = parameters.discretization
    num_lattices = parameters.num_lattices

    # Compute inverse coordinate
    i = 1 * constants.coords[component]
    j = 1 * constants.coords[component + num_directions]

    i_inv = -i
    j_inv = -j

    # Find inverse index
    inverse_component = constants.inverse_indices[component]

    # compute the shift to the target fluid lattice
    shift = get_index(i_inv, j_inv, dim=1)

    vector_self_index = (coordinate) + component * num_lattices
    vector_neightbour_index = (coordinate + shift) + inverse_component * num_lattices
    scalar_neightbour_index = coordinate + shift

    neightbour_velocity_x = velocity[scalar_neightbour_index]
    neightbour_velocity_y = velocity[scalar_neightbour_index + num_lattices]

    dot_product_UU = (neightbour_velocity_x * neightbour_velocity_x +
                      neightbour_velocity_y * neightbour_velocity_y)

    dot_product_CU = i * neightbour_velocity_x + j * neightbour_velocity_y
    dot_product_CU_inv = i_inv * neightbour_velocity_x + j_inv * neightbour_velocity_y


    velocity_expansion = (constants.const_1 * dot_product_CU +
                          constants.const_2 * dot_product_CU * dot_product_CU -
                          constants.const_3 * dot_product_UU +
                          1.0)

    velocity_expansion_inv = (constants.const_1 * dot_product_CU_inv +
                              constants.const_2 * dot_product_CU_inv * dot_product_CU_inv -
                              constants.const_3 * dot_product_UU +
                              1.0)


    # Compute equilibrium distribution
    equilibrium = (constants.weights[component] *
                   boundary_info.density_outflow *
                   velocity_expansion)

    equilibrium_inv = (constants.weights[inverse_component] *
                       boundary_info.density_outflow *
                       velocity_expansion_inv)

    population[vector_neightbour_index] = (equilibrium +
                                           equilibrium_inv -
                                           population[vector_self_index])

File: sequential-LBM/src/boundary_conditions.cpp
#include "headers/parameters.h"
#include "headers/helper.h"
#include "headers/boundary_conditions.h"

void TreatBoundary(ptr_boundary_func *boundary_update,
                   int *boundary_coords,
                   int num_boundaries,
                   real *population,
                   real *velocity,
                   real *density) {
    for (int component = 0; component < parameters.discretization; ++component) {
        for (int index = 0; index < num_boundaries; ++index) {
            int shift = component * num_boundaries;
            boundary_update[index + shift](component,
                                           boundary_coords[index],
                                           population,
                                           velocity,
                                           density);
        }
    }
}

void SkipBoundary(int component,
                  int coordinate,
                  real *population,
                  real *velocity,
                  real *density) {
}

void ApplyNonSlipBC(int component,
                    int coordinate,
                    real *population,
                    real *velocity,
                    real *density) {
    int num_lattices = parameters.num_lattices;
    int num_directions = parameters.discretization;

    int i = -1 * coords[component];
    int j = -1 * coords[component + num_directions];

    int inverse_component = inverse_indices[component];

    int shift = GetIndex(i, j);
    int self_index = coordinate + component * num_lattices;
    int neighbour_index = (coordinate + shift) + inverse_component * num_lattices;

    population[neighbour_index] = population[self_index];
}

void ApplyMovingWallBC(int component,
                       int coordinate,
                       real *population,
                       real *velocity,
                       real *density) {
    int num_lattices = parameters.num_lattices;
    int num_directions = parameters.discretization;

    int i = -1 * coords[component];
    int j = -1 * coords[component + num_directions];

    int inverse_component = inverse_indices[component];

    int shift = GetIndex(i, j);
    int self_index = coordinate + component * num_lattices;
    int neighbour_index = (coordinate + shift) + inverse_component * num_lattices;

    int scalar_neighbour_index = coordinate + shift;

    real const_one = constants.one;
    real const_two = constants.two;
    real const_three = constants.three;

    real dot_product_cu = const_one * (real(i) * boundary_info.wall_velocity_x +
                                       real(j) * boundary_info.wall_velocity_y);

    real complement = 2.0 * weights[component] * density[scalar_neighbour_index] * dot_product_cu;

    population[neighbour_index] = population[self_index] + complement;
}

void ApplyInflowBC(int component,
                   int coordinate,
                   real *population,
                   real *velocity,
                   real *density) {
    int num_lattices = parameters.num_lattices;
    int num_directions = parameters.discretization;

    int i = -1 * coords[component];
    int j = -1 * coords[component + num_directions];

    int inverse_component = inverse_indices[component];

    int shift = GetIndex(i, j);
    // int self_index = coordinate + component * num_lattices;
    int scalar_neighbour_index = coordinate + shift;
    int neighbour_index = scalar_neighbour_index + inverse_component * num_lattices;

    real dot_product_uu = boundary_info.velocity_inflow_x * boundary_info.velocity_inflow_x
                        + boundary_info.velocity_inflow_y * boundary_info.velocity_inflow_y;

    real dot_product_cu = i * boundary_info.velocity_inflow_x
                        + j * boundary_info.velocity_inflow_y;

    real const_one = constants.one;
    real const_two = constants.two;
    real const_three = constants.three;

    real velocity_expansion = const_one * dot_product_cu
                            + const_two * dot_product_cu * dot_product_cu
                            - const_three * dot_product_uu
                            + 1.0;

    real equilibrium = weights[inverse_component]
                     * density[scalar_neighbour_index]
                     * velocity_expansion;

    population[neighbour_index] = equilibrium;
}

void ApplyOutflowBC(int component,
                    int coordinate,
                    real *population,
                    real *velocity,
                    real *density) {
    int num_lattices = parameters.num_lattices;
    int num_directions = parameters.discretization;

    int i = -1 * coords[component];
    int j = -1 * coords[component + num_directions];

    int inverse_component = inverse_indices[component];

    int shift = GetIndex(i, j);

    int self_index = coordinate + component * num_lattices;

    int scalar_neighbour_index = coordinate + shift;
    int neighbour_index = scalar_neighbour_index + inverse_component * num_lattices;

    real neighbour_velocity_x = velocity[scalar_neighbour_index];
    real neighbour_velocity_y = velocity[scalar_neighbour_index + num_lattices];

    real dot_product_uu = neighbour_velocity_x * neighbour_velocity_x
                        + neighbour_velocity_y * neighbour_velocity_y;

    real dot_product_cu = -i * neighbour_velocity_x - j * neighbour_velocity_y;
    real dot_product_cu_inv = i * neighbour_velocity_x + j * neighbour_velocity_y;

    real const_one = constants.one;
    real const_two = constants.two;
    real const_three = constants.three;

    real velocity_expansion = (const_one * dot_product_cu)
                            + (const_two * dot_product_cu * dot_product_cu)
                            - (const_three * dot_product_uu)
                            + 1.0;

    real velocity_expansion_inv = (const_one * dot_product_cu_inv)
                                + (const_two * dot_product_cu_inv * dot_product_cu_inv)
                                - (const_three * dot_product_uu)
                                + 1.0;

    real equilibrium = weights[component]
                     * density[scalar_neighbour_index]
                     * velocity_expansion;

    real equilibrium_inv = weights[inverse_component]
                         * density[scalar_neighbour_index]
                         * velocity_expansion_inv;

    population[neighbour_index] = equilibrium
                                + equilibrium_inv
                                - population[self_index];
}
